Add codegen configuration file support

JoelSpeed opened the pull request:
This PR adds a new configuration file in the form of a per-group `.codegen.yaml` file.

The syntax is exemplified in the example API group, but I'll copy here for discussion:
```
schemapatch:
  enabled: true
  requiredFeatureSets:
  - ""
  - "Default"
  - "TechPreviewNoUpgrade"
swaggerdocs:
  enabled: true
  enforceComments: false
  outputFileName: zz_generated.swagger_doc_generated.go
compatibility:
  enabled: true
```

Changes of note:
* The schemapatch generator can now handle multiple feature set generations in a single run (via config file only)
* The swaggerdocs has a new `enforce-comments` option for use with `verify`, which is now configurable per group
* I've set up config files for all of the groups that use feature sets so far
* The config file, if found, will take precedence for configuration over the flags

ToDo:
* Get feedback on config design
* Make sure config overrides flags is what we want to do

JoelSpeed commented:
/test all

JoelSpeed commented:
/test verify

JoelSpeed commented:
Reviewers may want to start by reading through the documentation I've added, https://github.com/openshift/api/blob/92fed6d36abcbcfb805019dbace8cdd10bf18dfe/tools/codegen/README.md

deads2k commented on tools/codegen/README.md:
```diff
@@ -9,28 +9,24 @@ easier.
+  The generator will exit with a non-zero exit code if any file needs to be regenerated.
+- `--versions` - Prints the version of the utility and exits.
+
+When using the root command nakedly, all default generators will be executed on the API groups unless a configuration file
```
I like this behavior too.

deads2k commented on tools/codegen/README.md:
```diff
@@ -78,3 +74,100 @@ update-codegen-crds:
+For example, this annotation would be `+openshift:compatibility-gen:level=1` for a level 1 API.
+	
+Valid API levels are 1, 2, 3 and 4. Version 1 is required for all stable APIs.
+Version 2 should be used for beta level APIs. Levels 3 and 4 may be used for alpha APIs.
```
note for future us.  openshift/api should never again have a beta api version.  We're either devpreview and alpha, techpreview and alpha, or supported and stable.

JoelSpeed replied:
Yes I had noticed this as a tendency, I intend to add a section to the API conventions about when to use alpha vs stable APIs and will make a note in that about not using beta APIs

deads2k commented on tools/codegen/README.md:
```diff
@@ -78,3 +74,100 @@ update-codegen-crds:
+To generate CRD schemas, use the `schemapatch` subcommand.
+
+The schemapatch subcommand uses the [controller-tools][controller-tools] project to generate CRD schemas.
+It requires a stub CRD to exist before the schema will be generated. Copying an existing OpenShift CRD
```
another idea for future us, why bother with the stub.  An opt-out comment on a type containing typemeta and objectmeta seems more practical.

JoelSpeed replied:
I assume there is some historic reason for why we were using the schema patch instead of the full generator. I wonder if the full generation would clear the annotations that we set?

I could see a future version of this tool looking for a CRD yaml, if it doesn't exist, creating one, else, run the schema patch.

One irritation of the schemapatch tool is that it doesn't maintain customer printer columns, so that's another thing we can solve when we solve the stub issue

deads2k commented on tools/codegen/README.md:
```diff
@@ -78,3 +74,100 @@ update-codegen-crds:
+- `--controller-gen` - optionally use a particular controller-gen binary. When not specified, the tool will use the
+  built in generator.
+  Note, you must use a `controller-gen` built from the [OpenShift fork](https://github.com/openshift/kubernetes-sigs-controller-tools) of controller-tools.
+- `--required-feature-sets` - optionally generate based on the OpenShift feature sets annotations.
```
do we have an eventual goal to stop allowing this to be set?  Why only generate some of the schemas?

JoelSpeed replied:
I introduced this originally because of how the feature sets are being handled in the controller-gen tool, where it can only check against a single value for `OPENSHIFT_REQUIRED_FEATURESETS`.

I can see a future iteration where, when not specified, we work out which feature set combinations exist within the API group and then generate them all

deads2k commented on tools/codegen/README.md:
````diff
@@ -78,3 +74,100 @@ update-codegen-crds:
+```yaml
+# Configuration for the compatibility generator.
+compatibility:
+  enabled: true # Defaults to true, set to false to disable.
````
I haven't gotten to code yet, but I suggest inverting this to disabled so that your zero-value object matches your defaults.

deads2k replied:
pointer disabled is my preference.

elmiko replied:
> but I suggest inverting this to disabled so that your zero-value object matches your defaults.

this makes a lot of sense to me

deads2k commented on tools/codegen/README.md:
```diff
@@ -78,3 +74,100 @@ update-codegen-crds:
+# Configuration for the schemapatch generator.
+schemapatch:
+  enabled: true # Defaults to true, set to false to disable.
+  requiredFeatureSets: # Each entry will be matched against the value of the required feature set annotation.
```
What is the default?  I'd like to avoid having any package in openshift need to set this.

JoelSpeed replied:
The default is that it only generates manifests that don't have an annotation at all, which is the same behaviour as `controller-gen`. If we want to avoid users having to configure this, then we will have to add discovery of the required feature sets

We should collapse this conversation into https://github.com/openshift/api/pull/1343/files#r1025296179

deads2k commented on tools/codegen/README.md:
```diff
@@ -78,3 +74,100 @@ update-codegen-crds:
+# Configuration for the swaggerdocs generator.
+swaggerdocs:
+  enabled: true # Defaults to true, set to false to disable.
+  enforceComments: false # Defaults to false, set to true to enforce comments when verifying.
```
how many are we missing?  I accept that a generic tool needs this and we may even need this in our current state, but how many are broken?

deads2k replied:
eek.  Ok, we'll start here.

deads2k commented:
The direction of this tool for our repo looks good for people creating PRs to add or modify types and fields.  I have some comments about usage, the config file, and future refinements.  Future refinements are not required for merge.

The overall shape looks ok, I'm comfortable with whoever you think is best and available to do the review.

elmiko commented on tools/codegen/pkg/swaggerdocs/generator.go:
```diff
@@ -12,6 +12,16 @@ import (
 
 // Options contains the configuration required for the swaggerdocs generator.
 type Options struct {
+	// Enabled indicates whether the swaggerdocs generator is enabled or not.
```
i notice that we are using `Enabled` here in the comments which matches the value in the struct, but in the other locations (in this file and in `tools/codegen/pkg/generation/types.go`) we are using the serializable version of the name. i think for clarity these should be similar format, i'm not sure which the correct version though.

i notice in `tools/codegen/pkg/generation/generator.go` we use the name with the same capitalization in the comments as in the struct. perhaps these should all be capitalized the same?

JoelSpeed replied:
I've moved everything to the normal godocs style which matches the field names. I was thinking the config file should have the serializable names as the docs then match what the users would see, but we aren't going to generate docs from this so it's ok, they should be consistent now

elmiko replied:
makes sense, i was just curious because of the difference

elmiko reviewed (commented):
in general this PR seems good to me, i did have a question about some of the code comments, but the code seems fine to me (although i admit i'm not overly familiar with this code base). 

i tried testing this by checking out the branch and running `make update`, but it gave me an error because `GOPATH` was unset:

```
hack/update-deepcopy.sh                                                                                                                                                                                                                 make[1]: Entering directory '/home/mike/dev/openshift-api/tools'                                                                                                                                                                        go build -mod=vendor -o /home/mike/dev/openshift-api/tools/_output/bin/linux/amd64/deepcopy-gen ./vendor/k8s.io/code-generator/cmd/deepcopy-gen                                                                                         make[1]: Leaving directory '/home/mike/dev/openshift-api/tools'                                                                                                                                                                         
Generating Deepcopy for apiserver/v1 apps/v1 authorization/v1 build/v1 cloudnetwork/v1 config/v1 config/v1alpha1 console/v1 console/v1alpha1 example/v1 example/v1alpha1 helm/v1beta1 imageregistry/v1 image/v1 kubecontrolplane/v1 lega
cyconfig/v1 machine/v1 machine/v1alpha1 machine/v1beta1 monitoring/v1alpha1 networkoperator/v1 network/v1 oauth/v1 openshiftcontrolplane/v1 operatorcontrolplane/v1alpha1 operatoringress/v1 operator/v1 operator/v1alpha1 osin/v1 platf
orm/v1alpha1 project/v1 quota/v1 route/v1 samples/v1 securityinternal/v1 security/v1 servicecertsigner/v1alpha1 sharedresource/v1alpha1 template/v1 unidling/v1alpha1 user/v1 image/docker10 image/dockerpre012                         
hack/update-protobuf.sh                                                                                                                                                                                                                 
make[1]: Entering directory '/home/mike/dev/openshift-api/tools'                                                                                                                                                                        
go build -mod=vendor -o /home/mike/dev/openshift-api/tools/_output/bin/linux/amd64/go-to-protobuf ./vendor/k8s.io/code-generator/cmd/go-to-protobuf                                                                                     
make[1]: Leaving directory '/home/mike/dev/openshift-api/tools'                                                                                                                                                                         make[1]: Entering directory '/home/mike/dev/openshift-api/tools'                                                                                                                                                                        go build -mod=vendor -o /home/mike/dev/openshift-api/tools/_output/bin/linux/amd64/protoc-gen-gogo ./vendor/k8s.io/code-generator/cmd/go-to-protobuf/protoc-gen-gogo                                                                    
make[1]: Leaving directory '/home/mike/dev/openshift-api/tools'                                                                                                                                                                         
hack/update-protobuf.sh: line 32: GOPATH: unbound variable                                                                                                                                                                              
make: *** [Makefile:93: update-protobuf] Error 1                                                                                                                                                                                    ```

similar with `make verify`

```
+ /home/mike/dev/openshift-api/hack/update-protobuf.sh                                                              
make[1]: Entering directory '/home/mike/dev/openshift-api/tools'                                                    
make[1]: Leaving directory '/home/mike/dev/openshift-api/tools'                                                                                                                                                                         
make[1]: Entering directory '/home/mike/dev/openshift-api/tools'                                                    
make[1]: Leaving directory '/home/mike/dev/openshift-api/tools'                                                     
/home/mike/dev/openshift-api/hack/update-protobuf.sh: line 32: GOPATH: unbound variable                             
+ cleanup                                
+ rm -rf /home/mike/dev/openshift-api/_tmp                                                                          
make: *** [Makefile:44: verify-scripts] Error 1  
```

perhaps the update-codegen.sh script could check for `GOPATH` and give an error if not found, or perhaps try to default to `${HOME}/go` ?

fwiw, when i set `GOPATH` everything worked as expected.

JoelSpeed commented:
In the last commit, I've switch the `enforceComments` bool to a `commentPolicy` enum, I decided that it would be good to allow users outside of this repo to choose to ignore comments if they don't care about them, so now the values that are valid there are `Ignore`, `Warn` and `Strict`, defaulting to `Warn`.

Also switched to `disabled` as per the discussion.

> i tried testing this by checking out the branch and running make update, but it gave me an error because GOPATH was unset:

Looks like the protobuf script has an error, that isn't integrated into this tool yet, I'll check that out separately.

JoelSpeed commented:
@elmiko If you try `PROTO_OPTIONAL=true make update` outside of the GOPATH I think that should work. The protobuf generator doesn't work outside of the GOPATH right now. I have a [PR upstream](https://github.com/kubernetes/kubernetes/pull/112061) to make it work but it hasn't had any review yet.

JoelSpeed commented:
> In the last commit, I've switch the `enforceComments` bool to a `commentPolicy` enum, I decided that it would be good to allow users outside of this repo to choose to ignore comments if they don't care about them, so now the values that are valid there are `Ignore`, `Warn` and `Strict`, defaulting to `Warn`.
> 
> Also switched to `disabled` as per the discussion.
> 
> > i tried testing this by checking out the branch and running make update, but it gave me an error because GOPATH was unset:
> 
> Looks like the protobuf script has an error, that isn't integrated into this tool yet, I'll check that out separately.

Changed this to `Ignore`, `Warn` and `Enforce`, I think this PR is now ready to merge, I'm happy with the configuration API

elmiko commented on tools/codegen/pkg/generation/types.go:
```diff
@@ -0,0 +1,63 @@
+	// This generator is enabled by default so this field defaults to false.
+	Disabled bool `json:"disabled,omitempty"`
+
+	// requiredFeatureSets is a list of feature sets combinations that should be
```
```suggestion
	// RequiredFeatureSets is a list of feature sets combinations that should be
```
?

elmiko commented on tools/codegen/pkg/generation/types.go:
```diff
@@ -0,0 +1,63 @@
+	// When set to `Enforce`, the generator will return an error for any missing comments.
+	CommentPolicy string `json:"commentPolicy,omitempty"`
+
+	// outputFileName is the file name to use for writing the generated swagger
```
```suggestion
	// OutputFileName is the file name to use for writing the generated swagger
```

elmiko reviewed (changes requested):
just a couple missed comments

> If you try PROTO_OPTIONAL=true make update outside of the GOPATH I think that should work.

i tried this and it worked as expected

elmiko reviewed (commented):
/lgtm

thanks for the updates Joel

JoelSpeed commented:
/label px-approved
/label qe-approved
/label docs-approved
